developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+
+DSP: has a full topology, including host and guests' components. No distinction
+is made between the host and any guests.
+
+host: the SOF driver core, running on the host, and the ASoC subsystem only see
+the host topology. The vhost driver manages connection points between the host
+and the guests' topology fragments.
+
+guest: only "sees" the guest parts.
+
+As mentioned above, SOF IPC messages from guests are forwarded exactly 1-to-1 to
+the DSP. E.g., if a guest sends an IPC to adjust audio volume, it will be
+forwarded to the DSP. That IPC contains the volume component ID as seen by the
+respective guest. Therefore it should be exactly the same ID, as the one used by
+the firmware.
+
+3. Implementation
+*****************
+
+3.1. Initialisation
+-------------------
+
+The SOF driver probing on the host remains unmodified, except additionally the
+vhost driver is initialised, which registers a /dev/vhost-dsp misc device.
+
+When QEMU starts a guest instance, it detects the above misc device and
+creates a PCI device for the guest with dedicated vendor and device IDs.
+
+The SOF VirtIO driver on the guest registers support for that PCI device and
+its .probe() method is called.
+
+The SOF driver on the guest performs partial initialisation, it skips any
+steps, involving communicating with the actual DSP, e.g. sending the firmware,
+booting it, tracing initialisation.
+
+The VirtIO driver on the guest uses a new SOF_IPC_TPLG_VFE_GET VirtIO-specific
+IPC message to request the topology from the host. The host reads the respective
+file from its filesystem, and sends it in multiple chunks to the guest in reply
+to that request.
+
+TODO: define a reliable secure topology selection protocol. Currently the guest
+simply sends a configured ID to the host, upon which the host opens the topology
+file with the same name. Currently this isn't a problem, since so far only a
+single guest is supported. So, the host can verify, that the ID, that the guest
+has sent does indeed match with the guest topology file name. However, when
+support for multiple guests is added, it must be ascertained, that guests only
+receive topology files, that they are allowed to get. At the moment there
+doesn't seem to be a way on the host to uniquely and securely identify guests.
+If there is indeed no such way, our only possibility seems to be using
+cryptography for that.
+
+Guest topology files contain virtual DAI components of type "dai_out" and
+"dai_in." Their data tuples contain a new token of type SOF_TKN_COMP_REF, which
+contains an ID of the mixer or the demux, to which this pipeline should be
+connected.
+
+After obtaining the topology, the guest SOF instance uses a second new
+SOF_IPC_TPLG_VFE_COMP_ID VirtIO-specific IPC message to request its allocated
+component ID base. The host maintains a bitmap of guest component ID ranges to
+guarantee, that component IDs stay unique regardless of the order in which
+guests are brought up and down.
```

Comment:
I am confused by the mix of bitmap and ranges. What are you trying to say?

Reply:
well, I wasn't sure whether I had to put this here. It's already rather low level, but looks like I have to explain more.